plugins/SpectrumAnalyzer/Analyzer.cpp

```diff
@@ -50,17 +55,57 @@ extern "C" {
 Analyzer::Analyzer(Model *parent, const Plugin::Descriptor::SubPluginFeatures::Key *key) :
 	Effect(&analyzer_plugin_descriptor, parent, key),
 	m_processor(&m_controls),
-	m_controls(this)
+	m_controls(this),
+	m_processorThread(m_processor, m_inputBuffer, m_notifier),
+	// Buffer is sized to cover 4* the current maximum LMMS audio buffer size,
+	// so that it has some reserve space in case data processor is busy.
+	m_inputBuffer(4 * m_maxBufferSize)
 {
+	m_processorThread.start();
 }
 
 
+Analyzer::~Analyzer()
+{
+	m_processor.terminate();
+	m_notifier.wakeAll();
+	m_processorThread.wait();
+}
+
 // Take audio data and pass them to the spectrum processor.
-// Skip processing if the controls dialog isn't visible, it would only waste CPU cycles.
 bool Analyzer::processAudioBuffer(sampleFrame *buffer, const fpp_t frame_count)
 {
+	// Measure time spent in audio thread; both average and peak should be well under 1 ms.
+	#ifdef SA_DEBUG
+		unsigned int audio_time = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+		if (audio_time - m_last_dump_time > 5000000000)	// print every 5 seconds
+		{
+			std::cout << "Analyzer audio thread: " << m_sum_execution / m_dump_count << " ms avg / "
+				<< m_max_execution << " ms peak." << std::endl;
+			m_last_dump_time = audio_time;
+			m_sum_execution = m_max_execution = m_dump_count = 0;
+		}
+	#endif
+
 	if (!isEnabled() || !isRunning ()) {return false;}
-	if (m_controls.isViewVisible()) {m_processor.analyse(buffer, frame_count);}
+
+	// Skip processing if the controls dialog isn't visible, it would only waste CPU cycles.
+	if (m_controls.isViewVisible()) {
+		// To avoid processing spikes on audio thread, data are stored in
+		// a lockless ringbuffer and processed in a separate thread.
+		sampleFrame_copier copier(buffer);
+		m_inputBuffer.write_func<sampleFrame_copier>(copier, frame_count);
+
+		// Inform processor to check the buffer (to avoid busy waiting).
+		m_notifier.wakeAll();
```

Comment:
Could you explain how exactly does it cause a behavior that is not real-time safe?

Assuming you the effect processing thread has real-time priority (otherwise it does not matter), even if the `QWaitCondition` uses a mutex internally and it causes a syscall which returns control to the OS kernel, the kernel will just handle the syscall and return control back to your process (because you have real-time priority).

The way the mutex is used guarantees that it will always be free and available, because it only prevents multiple threads from trying to call the "wake up" at the same time. In this case there is only one thread such thread, so there is nothing else that could take the mutex, i.e. it will be always acquired instantly.

We also discussed delays that could occur because of context switch from user space to kernel space -- but these will occur anyway, since to achieve multi-tasking the kernel has to re-schedule what processes and threads are running every few milliseconds.

So, to summarize: syscall: no delay because you have realtime priority and context switches occur all the time anyway. Mutex: no delay because it is always free to take. So I still don't see the problem.

On the other hand, busy waiting has definite drawbacks: with high poll rate it needlessly burns CPU time (with cumulative effect over multiple plugin instances), and when the poll rate is low it wastes available CPU time by sleeping too long when you could have been processing data already.

Reply:
I'm really sorry, I've changed this text, but it seemed like my internet connection was broken at the time of editing.

I agree to all what you say, if you replace mutex by condition variable. Especially, I think you're right that this code does not hurt any of the 3 arguments not to use "mutexes" mentioned in the [article from Ross Bencina](http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing). So, in theory, we could whitelist waking up threads in condition variables in stoat.

In practice, however:

1. `QWaitCondition` [uses](https://code.woboq.org/qt5/qtbase/src/corelib/thread/qwaitcondition_unix.cpp.html) a `pthread_condition` variable and a `pthread_mutex_t`.
2. `pthread_cond_wait` [uses](https://code.woboq.org/userspace/glibc/nptl/pthread_cond_wait.c.html) `__pthread_cond_mutex_lock` (e.g. a mutex lock) plus a spinlock.

I imagine both cases are equally dangerous as the ones with the `sem_post`? @fundamental should we whitelist `pthread_cond_wait`, or do you consider it worse than `sem_post`?

Reply:
The programming patterns that pthread_cond_wait tend to have worse semantics than sem_post. sem_post does involve an internal futex lock, but that lock is virtually never blocked while pthread_cond_wait assumes that the section should fairly regularly block waiting for an event to occur.